collector/src/main.rs

```diff
@@ -397,15 +397,41 @@ fn get_local_toolchain(
     // + prefixed rustc is an indicator to fetch the rustc of the toolchain
     // specified. This follows the similar pattern used by rustup's binaries
     // (e.g., `rustc +stage1`).
-    let rustc = if rustc.starts_with('+') {
-        let s = String::from_utf8(
-            Command::new("rustup")
-                .args(&["which", "rustc", "--toolchain", &rustc[1..]])
-                .output()
+    let rustc = if let Some(toolchain) = rustc.strip_prefix('+') {
+        let output = Command::new("rustup")
+            .args(&["which", "rustc", "--toolchain", &toolchain])
+            .output()
+            .context("failed to run `rustup which rustc`")?;
+
+        // Looks like a commit hash? Try to install it...
+        if output.status.code() == Some(101) && toolchain.len() == 40 {
+            // No such toolchain exists, so let's try to install it with
+            // rustup-toolchain-install-master.
+
+            if !Command::new("rustup-toolchain-install-master")
```

Comment:
A common failure mode here would like be not having rustup-toolchain-install-master installed. Should we specifically handle that case?

Reply:
Yeah, we actually have sysroot building code built-in already and we should use that, I had forgotten about it when writing this.